language-design.md

```diff
@@ -38,16 +76,43 @@ you know a poorly designed language? What are the symptoms?
+
+We also think easy of reading is increasingly an attribute of well designed languages. Zoab feels that that the visual appearance of code should correspond to its logical structure. Alex somewhat agrees but also thinks there is a niche for languages that are easy to write at the expense of being hard to read (some shell scripting).
+
+Another plus is when a language _makes it easy to write what you intend, and hard to write what you don't intend_. That is, languages generally have a set of functionality they intend to support, and it would be nice if users could not accidentally do other things. [Rust](http://rust-lang.org) for example is a systems language which enforces memory safety. Given that most systems programmers don't want to violate memory safety (read uninitialized data or derefence null pointers), it's nice that the language prevents this.
+
+Documentation is also a great! :heart: Javadoc :heart:. It should always be unambiguous how a language handles a particular situation. As an example, [LOLCODE](http://lolcode.org/) will sometimes compile and sometimes not compile depending on how many PLZ's there are in your code (source: Dan Obermiller). While one could argue that requiring some numebr of PLZ's is not a good PL design choice, it is even worse to not precisely document what that requirement is.
 
+A poorly designed language, then, would be one lacking the good design paradigms that we have discussed above. For instance, LOLCODE can be considered a poorly designed language on account of its poorly documented compilation rules.
 
 ---
- 
+
+**Question** 
 
 In what way is an API a language? 
 
 **Response**
 
+We both agree that an API is indeed a language. To think about how an API is a language we note that languages have both vocabulary and syntax (both character choice and rules of composition). API's are very restricted in terms of syntax by their host language. This means that often their syntax is often trivial. That being said they still allow for a rich variety of vocabulary in the form of operations that can be performed.
+
+API's are also much esier to build than some type of DSLs (say external DSLs). For this reason many authors [Mernik et al., 2005] identify API's as being the first stage in the development of DSLs that ultimately become complex external DSLs.
+
+A number of concerns in designing good APIs (as identified by Bloch [Bloch, 2006]) are also concerns in designing a language.
+
+While one might be tempted to divide languages from APIs on the basis of syntax complexity, some API's end up having non-trivial syntax. For example, consider the SQL query generator [QueryDSL](https://github.com/querydsl/querydsl/tree/master/querydsl-sql). It allows the user to construct SQL queries using Java syntax. One might imagine that a query beginning
+
+    query_builder.from(my_table)
 
+might be well formed, but
+
+    query_builder.from(my_table).from(other_table)
```

Comment:
Daniel thinks this is allowed in SQL, but often does what you don't expect. He thinks this does a Cartesian join (all columns from both tables, with each row from one table paired with all rows from the other). Are you sure this isn't allowed in QueryDSL?